packages/toolpad-app/src/toolpad/AppEditor/PageEditor/QueriesExplorer/index.tsx

```diff
@@ -0,0 +1,502 @@
+import MoreVertIcon from '@mui/icons-material/MoreVert';
+import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
+import ChevronRightIcon from '@mui/icons-material/ChevronRight';
+import * as appDom from '../../../../appDom';
+import dataSources from '../../../../toolpadDataSources/client';
+import QueryIcon from '../../QueryIcon';
+import { useAppState, useAppStateApi } from '../../../AppState';
+import useBoolean from '../../../../utils/useBoolean';
+import EditableTreeItem, { EditableTreeItemProps } from '../../../../components/EditableTreeItem';
+import NodeMenu from '../../NodeMenu';
+import ExplorerHeader from '../../ExplorerHeader';
+
+const classes = {
+  treeItemMenuButton: 'Toolpad__QueryListItem',
+  treeItemMenuOpen: 'Toolpad__QueryListItemMenuOpen',
+};
+
+const StyledTreeItem = styled(EditableTreeItem)({
+  [`& .${classes.treeItemMenuButton}`]: {
+    visibility: 'hidden',
+  },
+  [`
+  & .${treeItemClasses.content}:hover .${classes.treeItemMenuButton},
+      & .${classes.treeItemMenuOpen}
+    `]: {
+    visibility: 'visible',
+  },
+});
+
+interface StyledTreeItemProps extends EditableTreeItemProps {
+  onDeleteNode?: (nodeId: NodeId) => void;
+  onDuplicateNode?: (nodeId: NodeId) => void;
+  onRenameNode?: (nodeId: NodeId, updatedName: string) => void;
+  onSelectNode?: (nodeId: NodeId) => void;
+  onCreate?: (event: React.MouseEvent, mode?: appDom.FetchMode) => void;
+  labelTextSx?: SxProps;
+  labelIconId?: string;
+  labelIconSx?: SxProps;
+  createLabelText?: string;
+  deleteLabelText?: string;
+  renameLabelText?: string;
+  duplicateLabelText?: string;
+  toolpadNodeId?: NodeId;
+}
+
+function DataTreeItem(props: StyledTreeItemProps) {
+  const {
+    nodeId,
+    labelText,
+    labelTextSx,
+    labelIconSx,
+    labelIconId,
+    onCreate,
+    onSelectNode,
+    onDeleteNode,
+    onDuplicateNode,
+    onRenameNode,
+    createLabelText,
+    deleteLabelText = 'Delete',
+    duplicateLabelText = 'Duplicate',
+    renameLabelText = 'Rename',
+    toolpadNodeId,
+    validateItemName,
+    ...other
+  } = props;
+
+  const { value: isEditing, setTrue: startEditing, setFalse: stopEditing } = useBoolean(false);
+
+  const handleRenameConfirm = React.useCallback(
+    (updatedName: string) => {
+      if (onRenameNode) {
+        onRenameNode(nodeId as NodeId, updatedName);
+        stopEditing();
+      }
+    },
+    [nodeId, onRenameNode, stopEditing],
+  );
+
+  const validateEditableQueryName = React.useCallback(
+    (newName: string) => {
+      if (newName !== labelText && validateItemName) {
+        return validateItemName(newName);
+      }
+      return { isValid: true };
+    },
+    [labelText, validateItemName],
+  );
+
+  const handleClick = React.useCallback(() => {
+    invariant(toolpadNodeId, 'DataTreeItem should only be used for nodes with a toolpadNodeId');
+    onSelectNode?.(toolpadNodeId);
+  }, [onSelectNode, toolpadNodeId]);
+
+  const queryCreationMode = React.useMemo(() => {
```

Comment:
> Ideally, the DataTreeItem doesn't know if it's a query or a mutation. In a true abstraction, we should be able to introduce a third query type without needing to alter DataTreeItem.

Agreed, naming it `DataTreeItem` was an attempt to begin making it agnostic to the type of item represented but I imagine that'll have to be a separate effort distinct from this one or any future improvements scoped to the query panel

> Also, I don't fully understand, do all tree nodes in the queries tree have the same node id? ids must be unique, otherwise we can't refer to the individual items in the treeview APIs. And in an ideal world they are opaque as well. (Meaning: we can't gain any knowledge about the object just by inspecting the id on its own)

The node ids are separate for each node